### PR TITLE
GH-18 Make OPA filter publish Spring event on deny

### LINKED ISSUE #18: Make filter use Spring Event bus
On `deny` filter should publish an appropriate event on the event bus.

### PR DESCRIPTION
**Pull request by irodzik**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import java.util.Base64
 
 allprojects {
-    version = "1.0.0-SNAPSHOT"
+    version = "2.0.0-SNAPSHOT"
```
Comment (anderseknert) [inline build.gradle.kts]: Should there not be a 1.0.0 before moving to 2.0.0-SNAPSHOT? Or there is one? :)

File: opa-filter-core/src/main/java/com/bisnode/opa/spring/security/filter/OpaFilter.java [hunk 2]
```diff
@@ -15,32 +18,37 @@
...
 
     /**
      * Creates new {@link OpaFilter} instance.
      *
-     * @param decider {@link AccessDecider} to use in {@link OpaFilter#doFilter(ServletRequest, ServletResponse, FilterChain)} method.
+     * @param decider        {@link AccessDecider} to use in {@link OpaFilter#doFilter(ServletRequest, ServletResponse, FilterChain)} method.
+     * @param eventPublisher {@link ApplicationEventPublisher} to publish events when deny occurs
      */
-    public OpaFilter(@NonNull AccessDecider<HttpServletRequest> decider) {
+    public OpaFilter(@NonNull AccessDecider<HttpServletRequest> decider, ApplicationEventPublisher eventPublisher) {
```
Comment (lukasz-kaminski) [inline opa-filter-core/src/main/java/com/bisnode/opa/spring/security/filter/OpaFilter.java:36]: I think we should additionally allow one-argument constructor, which would disable event publishing. Then, the version would be 1.1.0

**Review by lukasz-kaminski — CHANGES_REQUESTED** [review]
Can you, please, update readme?

File: opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy [hunk 3]
```diff
@@ -53,53 +61,70 @@ class OpaFilterConfigurerSpec extends Specification {
...
+          e.response.status == 403
+    }
+
+    def 'should publish AuthorizationFailureEvent on opa deny'() {
+        given:
+          wireMockServer.stubFor(any(anyUrl())
+                  .willReturn(aResponse()
+                          .withStatus(200)
+                          .withHeader(ContentType.HEADER_NAME, APPLICATION_JSON)
+                          .withBody('{"result": {"allow": false}}')
```
Comment (anderseknert) [inline opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy:85]: Do we have a test for OPA returning result without allow at all? Or just an empty response?

**Review by anderseknert — APPROVED** [review]
Great addition! I'm very much out of the Spring loop these days so I was not aware of spring events. Looks like a very nice value add.

File: opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy (hunk 3 above)
Comment (anderseknert) [reply opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy:85]: And, should events be published on error responses as well or that's meant to be handled by the usual exception handling only?

File: build.gradle.kts (hunk 1 above)
Comment (irodzik) [reply build.gradle.kts]: 1.0.0 is already released

File: opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy (hunk 3 above)
Comment (irodzik) [reply opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy:85]: Events are just additional way to inform developer that access was denied, so they can react to this in any way they want. In our case we are using it to check differences between OPA decision and our legacy solution decision.

Comment (irodzik) [reply opa-filter-autoconfigure/src/test/groovy/com/bisnode/opa/spring/security/filter/autoconfigure/OpaFilterConfigurerSpec.groovy:85]: > Do we have a test for OPA returning result without allow at all? Or just an empty response?

We don't, i will create test that will check that if we will get missing values we will deny.

File: opa-filter-core/src/main/java/com/bisnode/opa/spring/security/filter/OpaFilter.java (hunk 2 above)
Comment (irodzik) [reply opa-filter-core/src/main/java/com/bisnode/opa/spring/security/filter/OpaFilter.java:36]: Sounds good

File: build.gradle.kts (hunk 1 above)
Comment (anderseknert) [reply build.gradle.kts]: OK, shouldn't 1.0.0-SNAPSHOT come before 1.0.0? Not that it matters though, let's move on :)

File: README.md [hunk 4]
```diff
@@ -70,6 +70,8 @@ Unknown properties are ignored.
 > }
 > ```
 
+##Published events
```
Comment (lukasz-kaminski) [inline README.md]: You ate a space after `##`

image

**Review by lukasz-kaminski — APPROVED** [review]

**Pull request merged**